### ISSUE TITLE
 SerializeAsync System.Linq.Enumerable.SelectListIterator to stream error

### ISSUE THREAD

**Issue by IKende** [Tornhoof/SpanJson#115]
at SpanJson.JsonWriter`1.ThrowArgumentException(String message, String paramName) in C:\projects\spanjson\SpanJson\JsonWriter.cs:line 379
   at SpanJson.JsonWriter`1.WriteUtf8Double(Double value) in C:\projects\spanjson\SpanJson\JsonWriter.Utf8.cs:line 150
   at lambda_method(Closure , JsonWriter`1& , <>f__AnonymousType10`11 )
   at SpanJson.Formatters.ComplexClassFormatter`3.Serialize(JsonWriter`1& writer, T value) in C:\projects\spanjson\SpanJson\Formatters\ComplexClassFormatter.cs:line 33
   at SpanJson.Formatters.BaseFormatter.SerializeRuntimeDecisionInternal[T,TSymbol,TResolver](JsonWriter`1& writer, T value, IJsonFormatter`2 formatter) in C:\projects\spanjson\SpanJson\Formatters\BaseFormatter.cs:line 40
   at SpanJson.Formatters.EnumerableFormatter`4.Serialize(JsonWriter`1& writer, TEnumerable value) in C:\projects\spanjson\SpanJson\Formatters\EnumerableFormatter.cs:line 57
   at SpanJson.Formatters.RuntimeFormatter`2.Serialize(JsonWriter`1& writer, Object value) in C:\projects\spanjson\SpanJson\Formatters\RuntimeFormatter.cs:line 35
   at lambda_method(Closure , JsonWriter`1& , ActionResult )
   at SpanJson.Formatters.ComplexClassFormatter`3.Serialize(JsonWriter`1& writer, T value) in C:\projects\spanjson\SpanJson\Formatters\ComplexClassFormatter.cs:line 33
   at lambda_method(Closure , Object , Stream , CancellationToken )
   at SpanJson.JsonSerializer.NonGeneric.Inner`2.InnerSerializeAsync(Object input, Stream stream, CancellationToken cancellationToken) in C:\projects\spanjson\SpanJson\JsonSerializer.NonGeneric.cs:line 120
   at SpanJson.JsonSerializer.NonGeneric.Utf8.SerializeAsync(Object input, Stream stream, CancellationToken cancellationToken) in C:\projects\spanjson\SpanJson\JsonSerializer.NonGeneric.cs:line 521

the object can serialize in Newtonsoft.Json

**Comment by Tornhoof** [comment]
Serialization of LINQ datatypes is not supported.

**Comment by IKende** [comment]
SpanJson version is 2.0.7,
linq select properties are value type
```
            return from a in items
                   select new
                   {
                       a.ID,
                       a.Host,
                       a.Category,
                       a.Remark,
                       a.Available,
                       a.Status?.Cpu,
                       a.Status?.RequestPer,
                       a.Status?.TotalRequest,
                       Memory = a.Status?.Memory / 1024,
                       a.Status?.RunTime,
                       a.Gateway
                   };
```

**Comment by Tornhoof** [comment]
The line offsets in the exception are very weird. the only code which throws there is if the number is not a valid double for serialization (like NAN), but you cut off the message of the exception so it's hard to tell.

**Comment by IKende** [comment]
Not supported Nullable<double>?
```
<>f__AnonymousType7<long, string, string, bool, double?, long?, double?, double?, string>>
```
exception detail
```
 Data: {System.Collections.ListDictionaryInternal}
    HResult: -2147024809
    HelpLink: null
    InnerException: null
    Message: "Invalid double value for JSON\r\nParameter name: value"
    ParamName: "value"
    Source: "SpanJson"
    StackTrace: "   at SpanJson.JsonWriter`1.ThrowArgumentException(String message, String paramName) in C:\\projects\\spanjson\\SpanJson\\JsonWriter.cs:line 379\r\n   at SpanJson.JsonWriter`1.WriteUtf8Double(Double value) in C:\\projects\\spanjson\\SpanJson\\JsonWriter.Utf8.cs:line 150\r\n   at lambda_method(Closure , JsonWriter`1& , <>f__AnonymousType7`9 )\r\n   at SpanJson.Formatters.ComplexClassFormatter`3.Serialize(JsonWriter`1& writer, T value) in C:\\projects\\spanjson\\SpanJson\\Formatters\\ComplexClassFormatter.cs:line 33\r\n   at SpanJson.Formatters.BaseFormatter.SerializeRuntimeDecisionInternal[T,TSymbol,TResolver](JsonWriter`1& writer, T value, IJsonFormatter`2 formatter) in C:\\projects\\spanjson\\SpanJson\\Formatters\\BaseFormatter.cs:line 40\r\n   at SpanJson.Formatters.EnumerableFormatter`4.Serialize(JsonWriter`1& writer, TEnumerable value) in C:\\projects\\spanjson\\SpanJson\\Formatters\\EnumerableFormatter.cs:line 57\r\n   at SpanJson.Formatters.RuntimeFormatter`2.Serialize(JsonWriter`1& writ
er, Object value) in C:\\projects\\spanjson\\SpanJson\\Formatters\\RuntimeFormatter.cs:line 35\r\n   at lambda_method(Closure , JsonWriter`1& , ActionResult )\r\n   at SpanJson.Formatters.ComplexClassFormatter`3.Serialize(JsonWriter`1& writer, T value) in C:\\projects\\spanjson\\SpanJson\\Formatters\\ComplexClassFormatter.cs:line 33\r\n   at lambda_method(Closure , Object , Stream , CancellationToken )\r\n   at SpanJson.JsonSerializer.NonGeneric.Inner`2.InnerSerializeAsync(Object input, Stream stream, CancellationToken cancellationToken) in C:\\projects\\spanjson\\SpanJson\\JsonSerializer.NonGeneric.cs:line 120\r\n   at SpanJson.JsonSerializer.NonGeneric.Utf8.SerializeAsync(Object input, Stream stream, CancellationToken cancellationToken) in C:\\projects\\spanjson\\SpanJson\\JsonSerializer.NonGeneric.cs:line 521\r\n   at BeetleX.FastHttpApi.HttpApiServer.FrameSerialize(DataFrame data, Object body) in D:\\VisualStudio\\BeetleX\\BeetleX\\BeetleX.FastHttpAPI\\HttpApiServer.cs:line 467\r\n   at BeetleX.FastHttpApi
.WebSockets.DataFrame.BeetleX.FastHttpApi.IDataResponse.Write(PipeStream stream) in D:\\VisualStudio\\BeetleX\\BeetleX\\BeetleX.FastHttpAPI\\WebSockets\\DataFrame.cs:line 195\r\n   at BeetleX.FastHttpApi.HttpPacket.OnEncode(ISession session, Object data, Stream stream) in D:\\VisualStudio\\BeetleX\\BeetleX\\BeetleX.FastHttpAPI\\HttpPacket.cs:line 177\r\n   at BeetleX.FastHttpApi.HttpPacket.Encode(Object data, ISession session, Stream stream) in D:\\VisualStudio\\BeetleX\\BeetleX\\BeetleX.FastHttpAPI\\HttpPacket.cs:line 168\r\n   at BeetleX.TcpSession.WriterData(Object data, Stream stream) in D:\\VisualStudio\\BeetleX\\BeetleX\\src\\BeetleX\\TcpSession.cs:line 375\r\n   at BeetleX.TcpSession.ProcessSendMessages() in D:\\VisualStudio\\BeetleX\\BeetleX\\src\\BeetleX\\TcpSession.cs:line 294\r\n   at BeetleX.TcpSession.Send(Object data) in D:\\VisualStudio\\BeetleX\\BeetleX\\src\\BeetleX\\TcpSession.cs:line 357\r\n   at BeetleX.FastHttpApi.WebSockets.DataFrame.Send(ISession session) in D:\\VisualStudio\\BeetleX\\Be
etleX\\BeetleX.FastHttpAPI\\WebSockets\\DataFrame.cs:line 253\r\n   at BeetleX.FastHttpApi.WSActionResultHandler.Success(Object result) in D:\\VisualStudio\\BeetleX\\BeetleX\\BeetleX.FastHttpAPI\\IActionResultHandler.cs:line 65\r\n   at BeetleX.FastHttpApi.ActionContext.OnExecute(IActionResultHandler resultHandler) in D:\\VisualStudio\\BeetleX\\BeetleX\\BeetleX.FastHttpAPI\\ActionContext.cs:line 74"
    TargetSite: {Void ThrowArgumentException(System.String, System.String)
```

**Comment by Tornhoof** [comment]
No, it supports nullable, but I think somethign weird is happening there with that anonymous type, I'll take a closer look if I have time.

**Comment by Tornhoof** [comment]
This should be solved in 2.0.8. The check in the writer code was invalid, instead of just checking that the value is not NAN or infinite, we disallowed 0.0 too.
